Code changes for OSW 0.3

wisamuw opened the pull request:
- 3 new entities demonstrating adjacent entities
- Custom entities

wisamuw committed:
Code changes for 0.3:
- 3 new entities demonstratong adjacent entities
- Custom entities

uw-amy-bordenave commented:
@wisamuw 
Note, the `=mixed` value for the keys `leaf_type=` and `leaf_cycle=` is not valid on `natural=tree` nodes, only on `natural=tree_row` ways and `natural=wood` areas.

Ref: Current implementation, `fields.ts`, starting [line 202](https://github.com/OpenSidewalks/OpenSidewalks-Schema/blob/892738857c33eeb9a47277b1d4814e3f263ce6df/jsonschema/src/fields.ts#L202C1-L218C11)
```
/**
 * A field that describes the type of leaves, for instance broadleaved, or needleleaved.
 */
export type LeafType =
| "broadleaved"
| "needleleaved"
| "mixed"
| "leafless";
/**
 * A field that describes the phenology of leaves, for instance evergreen or deciduous.
 */
export type LeafCycle =
| "evergreen"
| "deciduous"
| "semi_evergreen"
| "semi_deciduous"
| "mixed";
```

---

Additionally, the evergreen/deciduous distinction can be useful for answering, "Does this tree provide shade in both winter and summer?", but I don't think `=semi_evergreen` and `=semi_deciduous` are really in-scope for OSW.

wisamuw committed:
- Bump schema version from 0.2 to 0.3
- Introduce the concept of subschemas to prevent entities from leaking across files
- Eliminate maximum value for length field
- Introduce the concept of compatible schemas
- Define 6 distinct custom entities for each file
- Replace absolute links within the schema with relative ones
- Address: Note, the =mixed value for the keys leaf_type= and leaf_cycle= is not valid on natural=tree nodes, only on natural=tree_row ways and natural=wood areas

uw-amy-bordenave commented:
Re-posting as requested by @wisamuw:

The evergreen/deciduous distinction can be useful for answering, "Does this tree provide shade in both winter and summer?", but I don't think `=semi_evergreen` and `=semi_deciduous` are really in-scope for OSW.

jeffmaki commented on jsonschema/src/polygons/wood.ts:
```diff
@@ -0,0 +1,29 @@
+ * Fields that identify woods.
+ */
+interface WoodIdentifyingFields extends BasePolygonFields {
+  natural: "wood";
```
Is "wood" here deliberate vs. "woods"? I think if you mean a wooded area, I'd keep as "woods" (my opinion ofc).

uw-amy-bordenave replied:
[`natural=wood`](https://wiki.openstreetmap.org/wiki/Tag:natural%3Dwood) is the OSM tag for a forest or "woods".

jeffmaki commented on opensidewalks.schema.json:
```diff
@@ -4572,86 +5696,48 @@
+        "TrunkRoad": {
             "additionalProperties": false,
-            "description": "The centerline of a footpath traversing a traffic island. Some complex, long, or busy pedestrian crossings have a built-up \"island\" to protect pedestrians, splitting up the crossing of the street into two or more crossings. As a pedestrian uses this crossing, they will transition across these path elements: sidewalk → footway → crossing → traffic island → crossing →  ...
+            "description": "A high-performance or high-importance roads that don't meet the requirements for motorway, but are not classified as highway=primary either.",
```
aka "arterial"?

wisamuw replied:
Yes, they are a subset of arterials that usually have higher speeds and are not pedestrian friendly, I think we only have one in Seattle (50mph max speed). We simply used OSM's definition :)

jeffmaki commented on opensidewalks.schema.json:
```diff
@@ -2949,18 +3875,25 @@
+                },
+                "width": {
+                    "description": "A field for width of an entity in meters.",
+                    "maximum": 500,
```
Is 500m just a sanity check? How'd we arrive at this?

wisamuw replied:
Yes, just a sanity check. Wanted a width that's unreasonably large as a ceiling. This has not changed since v0.2.

jeffmaki commented on README.md:
```diff
@@ -784,6 +802,58 @@ Polygons describe 2-dimensional areas which are adjacent to pedestrian paths. Th
 
 </details>
 
+## List of Custom Entities
```
Can users create more than one of these and give them names? If so, should there be a "name" or "type" attribute that isn't ext: prefixed?

cyrossignol replied:
The definitions of custom entities exist entirely outside of OSW. Producers can create as many of these as they need. We're just providing a container that enables these to pass validation. From OSW's point of view, any `ext:*`-only entity is custom.

uw-amy-bordenave committed:
Merge branch 'OSW0.3' into OSW0.3-code

uw-amy-bordenave commented:
Note: https://github.com/OpenSidewalks/OpenSidewalks-Schema/pull/21/commits/6a93bab0fe06f8c763d74a7ae0b9b03b1db35c38 just resolves a conflict in the main readme, by moving the `natural=wood` definition to the correct place. No content or code changes.

Edit: Not entirely sure what happened here. `https://github.com/OpenSidewalks/OpenSidewalks-Schema/pull/21/changes/6a93bab0fe06f8c763d74a7ae0b9b03b1db35c38` shows only changes to the readme, but `https://github.com/OpenSidewalks/OpenSidewalks-Schema/commit/6a93bab0fe06f8c763d74a7ae0b9b03b1db35c38` shows changes to `resources/josm/presets/opensidewalks.xml` and `resources/josm/presets/opensidewalks.xml`.
I'm going to blame GitHub UI on that one, as there was no indication that the "Resolve conflicts" menu also made other changes... @wisamuw feel free to revert https://github.com/OpenSidewalks/OpenSidewalks-Schema/pull/21/commits/6a93bab0fe06f8c763d74a7ae0b9b03b1db35c38 if this is causing issues.

cyrossignol commented:
> Edit: Not entirely sure what happened here. 

@uw-amy-bordenave 6a93bab is a merge commit and captures all of the differences at that point between `OSW0.3` and `OSW0.3-code`. Looks like `OSW0.3-code` was behind `OSW0.3 ` with those preset/mapcss changes, and now it's not. Since we're merging back into `OSW0.3`, and the section for `natural=wood` created a conflict, GitHub displays that diff in PR mode, but not the others from the merge that will apply cleanly. We don't need to revert the merge.

cyrossignol commented on jsonschema/src/generate.ts:
```diff
@@ -43,11 +43,46 @@ function addPatternProperties(schema: any) {
+  ];
+
+  const subSchema =
+    candidates.map((t) => TJS.generateSchema(subProgram, t, settings)).find(Boolean) as any;
```
Should each subschema have a different `$id`? Something like: 

```
https://sidewalks.washington.edu/opensidewalks/0.3/schema.nodes.json
https://sidewalks.washington.edu/opensidewalks/0.3/schema.edges.json
...
```

wisamuw replied:
As discussed in the schema meeting today. we see no value in having a unique identifier for each file type since our schema already relies on file naming convention to identify the file type.

wisamuw committed:
Addressing: I don't think =semi_evergreen and =semi_deciduous are really in-scope for OSW
